### PR TITLE
Add support for Platform Load Sharing Facility (LSF) backend.

### PR DESCRIPTION
**Pull request by sleongmgi**

### CODE REVIEW

File: engine/src/main/scala/cromwell/engine/backend/lsf/LsfBackend.scala [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+      s"""#!/bin/sh
+         |$instantiatedCommand
+         |echo $$? > rc
+         |""".stripMargin)
+  }
+
+  /**
+   * Launches the bsub command, returns a tuple: (rc, Option(lsf_job_id))
+   */
+  private def launchQsub(jobDescriptor: BackendCallJobDescriptor): (Int, Option[Int]) = {
```
Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/engine/backend/lsf/LsfBackend.scala]: Probably should be `launchBsub`?

**Comment by cjllanwarne** [comment]
@sleongmgi this is very cool! I believe based off of the SGE backend?

Have you had a chance to see what's happening on our Develop branch? We'll soon have a new engine-to-backend interface and it'd be awesome if we could also bring this backend forward with us!

In any case, thanks for the contribution! :-D

File: engine/src/main/scala/cromwell/engine/backend/JobDescriptor.scala [hunk 2]
```diff
@@ -17,7 +17,7 @@ import scala.util.Try
   *
   * @tparam K CallKey subtype
   */
-sealed trait JobDescriptor[K <: JobKey] {
+sealed trait JobDescriptor[K <: CallKey] {
```
Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/engine/backend/JobDescriptor.scala]: Out of interest, what prompted this change?

Comment (sleongmgi) [inline engine/src/main/scala/cromwell/engine/backend/JobDescriptor.scala]: Without the change it gave me an error when compiling.

File: engine/src/main/scala/cromwell/engine/backend/lsf/LsfBackend.scala (hunk 1 above)
Comment (sleongmgi) [inline engine/src/main/scala/cromwell/engine/backend/lsf/LsfBackend.scala]: Yes, I should call it `launchBsub`.

File: engine/src/main/scala/cromwell/engine/backend/JobDescriptor.scala (hunk 2 above)
Comment (sleongmgi) [inline engine/src/main/scala/cromwell/engine/backend/JobDescriptor.scala]: Yes, the LSF backend is based off of SGE backend.

I have not checked out the "Develop" branch yet.

Comment (sleongmgi) [inline engine/src/main/scala/cromwell/engine/backend/JobDescriptor.scala]: I just checked the "develop" branch.  It is very cool.  I will give it a try to put the LSF as a supportedBackends.

Comment (geoffjentry) [inline engine/src/main/scala/cromwell/engine/backend/JobDescriptor.scala]: @sleongmgi If you send me your email address (either here or email me - my address is on my profile) I can send you a doc which goes into more detail on the new pluggable backend architecture which might be helpful

Comment (sleongmgi) [inline engine/src/main/scala/cromwell/engine/backend/JobDescriptor.scala]: That would be great.

File: .dockerignore [hunk 3]
```diff
@@ -1 +0,0 @@
-.gitignore
```
Comment (geoffjentry) [inline .dockerignore]: Can you explain what prompted this change? In particular removing the .gitignore

File: README.md [hunk 4]
```diff
@@ -325,14 +327,16 @@ The backend/filesystem pairings are as follows:
 
 * [Local Backend](#local-backend) uses the [Shared Local Filesystem](#shared-local-filesystem)
 * [SGE Backend](#sun-gridengine-backend) uses the [Shared Local Filesystem](#shared-local-filesystem)
+* [LSF Backend](#platform-load-sharing-facility-backend) uses the [Shared Local Filesystem](#shared-local-filesystem)
 * [JES Backend](#google-jes-backend) uses the [Google Cloud Storage Filesystem](#google-cloud-storage-filesystem)
 
 ### Shared Local Filesystem
 
-For the [local](#local-backend) and [Sun GridEngine](#sun-gridengine-backend) backends, the following is required of the underlying filesystem:
+For the [local](#local-backend) and [Sun GridEngine](#sun-gridengine-backend) and [Platform Sharing Facility](#platform-load-sharing-facility-backend) backends, the following is required of the underlying filesystem:
```
Comment (geoffjentry) [inline README.md]: Change the first 'and' to a comma

File: engine/src/main/scala/cromwell/engine/backend/lsf/LsfBackend.scala [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+         |echo $$? > rc
+         |""".stripMargin)
+  }
+
+  /**
+   * Launches the bsub command, returns a tuple: (rc, Option(lsf_job_id))
+   */
+  private def launchBsub(jobDescriptor: BackendCallJobDescriptor): (Int, Option[Int]) = {
+    val logger = jobLogger(jobDescriptor)
+    val lsfOption = new java.util.HashMap[String,Object]()
```
Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/engine/backend/lsf/LsfBackend.scala]: I actually think I preferred the original way of doing this (although I'm sure the assignment could have been generalised into a function).

In particular, I don't like the `java.util.HashMap` here. I personally am not used to mutable collections in the codebase and this one took me by surprise!

**Comment by cjllanwarne** [comment]
I think this'd be a great contribution. 👍 from me!

**Comment by geoffjentry** [comment]
@sleongmgi Can you take a look at the review feedback and then we can merge this in. Thanks!

**Comment by sleongmgi** [comment]
Thank you very much for all the feedbacks.  I will go through them and make necessary changes.

File: .dockerignore (hunk 3 above)
Comment (sleongmgi) [inline .dockerignore]: Yes, you are right.  This file should not have been changed.  I will restore back this.

**Comment by geoffjentry** [comment]
Awesome, thanks @sleongmgi !

**Pull request merged**